### PR TITLE
dune: init at 2.1.0

### PR DESCRIPTION
**Pull request by marsam**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by marsam** [comment]
@GrahamcOfBorg fstar

**Comment by marsam** [comment]
@GrahamcOfBorg build fstar

**Comment by vyorkin** [comment]
Do you want to add `dune-configurator` as well or it would be better to add it in a separate PR?

File: pkgs/development/tools/ocaml/cppo/default.nix [hunk 1]
```diff
@@ -7,11 +7,15 @@ assert stdenv.lib.versionAtLeast ocaml.version "3.12";
...
+    version = "1.6.6";
+    sha256 = "1smcc0l6fh2n0y6bp96c69j5nw755jja99w0b206wx3yb2m4w2hs";
     buildInputs = [ dune ];
     extra = {
-      inherit (dune) installPhase;
+      installPhase = ''
+        runHook preInstall
+        ${dune}/bin/dune install --prefix=$out --libdir=$OCAMLFIND_DESTDIR
+        runHook postInstall
+      '';
```
Comment (Zimmi48) [inline pkgs/development/tools/ocaml/cppo/default.nix]: I don't understand these changes. What are they meant to fix? If running `dune install` is generally better than running `opaline`, why not change this in the Dune derivation itself?

**Review by Zimmi48 — COMMENTED** [review]
Not having Dune 2.0.0 yet in nixpkgs is an inconvenience. Can you please explain what are the things that are blocking this PR or what additional work you think is required?

I am not sure I understand the first commit, adding Dune 2.0, but keeping Dune 1.11 without exposing it. Why would we want to keep Dune 1.11 in nixpkgs?

If this is because Dune 2.0 is not compatible with OCaml < 4.06, note that we could go back to the situation preceding #57953 where Dune was out of `ocamlPackages`. This was changed because Dune is both a tool and a library, but we could try to find another solution, e.g. making it available as a library only for `ocamlPackages` sets starting with 4.06 and as a tool independent of `ocamlPackages` for any OCaml version.

I suppose it is not because some OCaml packages have not been migrated yet from jbuilder to Dune, otherwise you wouldn't be updating a bunch of packages in this PR as well. BTW, for packages that still haven't migrated to Dune in their latest version, maybe that says something about them (not) being maintained, and we could simply choose to drop them, or at least mark them as broken and not care much...

File: pkgs/development/ocaml-modules/stdint/default.nix [hunk 2]
```diff
@@ -13,8 +13,20 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ ocaml findlib dune ];
 
+  patches = [
+    # Fetch dune patch @mor1 fork. TODO(marsam): create a PR in upstream
```
[the remark this answers is not in the document]
Comment (marsam) [reply pkgs/development/ocaml-modules/stdint/default.nix]: Thanks, fixed

File: pkgs/development/tools/ocaml/cppo/default.nix (hunk 1 above)
Comment (marsam) [reply pkgs/development/tools/ocaml/cppo/default.nix]: I simply prefer to use dune install over opaline. I recall trying to set it to into `buildDunePackage` but there were some failing packages, but I haven't investigated much. 
I'll try to set into `buildDunePackage` it in another PR

**Comment by marsam** [comment]
>Why would we want to keep Dune 1.11 in nixpkgs?

dune 2 is a major version bump, a there are multiple packages which still haven't been migrated, for instance, janestreet packages.

I tried to set dune 2 as default, but I encountered various errors, hence only added dune_2 only as another attribute. 

Only the last three commits are relevant; @vbgl should I cut the other commits into a separated PR?

**Comment by Zimmi48** [comment]
> > 
> 
> dune 2 is a major version bump, a there are multiple packages which still haven't been migrated, for instance, janestreet packages.

In theory, Dune 2 shouldn't break Dune projects which set the Dune language version to a lower version than 2.0. However, it does break the projects that have not migrated to Dune from jbuilder. Did you find examples beyond this? If yes, it could be interesting to look into them more closely and possibly report the issues to Dune developers.

In any case, this PR LGTM (in full, including the version bumps) and I'm in favor of merging it as-is now (without splitting).

**Comment by Zimmi48** [comment]
@marsam I'd like to encourage you to go ahead and merge this without waiting further.

**Comment by Zimmi48** [comment]
@marsam or @vbgl: please merge this now.

BTW: PR title should be edited to reflect the change (dune_2: init at 2.1.0).

**Comment by marsam** [comment]
@Zimmi48 sorry for the delay, I was waiting for another review; merging as-is because it shouldn't introduce any major change. @vbgl I hope you don't mind

**Comment by Zimmi48** [comment]
Thanks!

**Comment by Shados** [comment]
@marsam This PR included an update for cppo from 1.6.5 to 1.6.6. This broke building cppo with ocaml 4.02, which is still needed for some things.

**Comment by marsam** [comment]
@Shados sorry about that, reverted in https://github.com/NixOS/nixpkgs/pull/76942

**Pull request merged**